File: src/lib/electionTally.ts
export type ElectionPost = 'president' | 'vice_president' | 'secretary' | 'treasurer' | 'committee';

export type PollOptionRow = {
  id: string;
  poll_id: string | null;
  option_text: string;
  election_post: string | null;
  member_id?: string | null;
  flat_id?: string | null;
  flat_number?: string | null;
  nominated_by?: string | null;
  nomination_statement?: string | null;
};

/**
 * Bye-law ballot payload (non-ranked).
 * - Combined (default): `{ selected: optionId[] }` — mark up to committeeSize nominees.
 * - Separate office (expressly approved): `{ [post]: optionId }` and `{ committee: optionId[] }`.
 */
export type BallotChoices = {
  selected?: string[];
  president?: string;
  vice_president?: string;
  secretary?: string;
  treasurer?: string;
  committee?: string | string[];
};

export type BallotRow = {
  choices?: BallotChoices | Record<string, unknown> | null;
  /** @deprecated Legacy ranked/Borda ballots only. */
  rankings?: Record<string, Record<string, number>> | null;
};

const OFFICE_POSTS: ElectionPost[] = ['president', 'vice_president', 'secretary', 'treasurer'];
const ALL_POSTS: ElectionPost[] = [...OFFICE_POSTS, 'committee'];

function asStringArray(v: unknown): string[] {
  if (Array.isArray(v)) return v.filter((x): x is string => typeof x === 'string' && x.length > 0);
  if (typeof v === 'string' && v.length > 0) return [v];
  return [];
}

function normalizeChoices(raw: BallotRow['choices']): BallotChoices {
  if (!raw || typeof raw !== 'object') return {};
  const c = raw as BallotChoices;
  return {
    selected: asStringArray(c.selected),
    president: typeof c.president === 'string' ? c.president : undefined,
    vice_president: typeof c.vice_president === 'string' ? c.vice_president : undefined,
    secretary: typeof c.secretary === 'string' ? c.secretary : undefined,
    treasurer: typeof c.treasurer === 'string' ? c.treasurer : undefined,
    committee: asStringArray(c.committee),
  };
}

function postsPresent(options: PollOptionRow[]): ElectionPost[] {
  const have = new Set(options.map((o) => o.election_post).filter(Boolean));
  return ALL_POSTS.filter((p) => have.has(p));
}

/** Validate one-member ballot choices (plurality / limited vote — never ranked). */
export function validateElectionChoices(
  options: PollOptionRow[],
  choicesRaw: BallotChoices,
  separateOfficeVotes: boolean,
  opts?: { committeeSeats?: number; maxMarks?: number },
): string | null {
  const committeeSeats = opts?.committeeSeats ?? 3;
  const maxMarks = opts?.maxMarks ?? 7;
  const choices = normalizeChoices(choicesRaw);
  const byId = new Map(options.map((o) => [o.id, o]));
  const openPosts = postsPresent(options);

  if (separateOfficeVotes) {
    for (const post of openPosts) {
      if (post === 'committee') {
        const picks = asStringArray(choices.committee);
        if (picks.length === 0 && options.some((o) => o.election_post === 'committee')) {
          return 'Select Executive Member candidate(s)';
        }
        if (picks.length > committeeSeats) {
          return `Select at most ${committeeSeats} Executive Member(s)`;
        }
        const seen = new Set<string>();
        for (const id of picks) {
          const opt = byId.get(id);
          if (!opt || opt.election_post !== 'committee') return 'Invalid Executive Member selection';
          if (seen.has(id)) return 'Duplicate Executive Member selection';
          seen.add(id);
        }
        continue;
      }
      const pick = choices[post];
      if (!pick) return `Select one candidate for ${post.replace('_', ' ')}`;
      const opt = byId.get(pick);
      if (!opt || opt.election_post !== post) return `Invalid selection for ${post.replace('_', ' ')}`;
    }
    return null;
  }

  const selected = asStringArray(choices.selected);
  if (selected.length === 0) return 'Select at least one nominee on your ballot';
  if (selected.length > maxMarks) return `Select at most ${maxMarks} nominees (one vote per member)`;
  const seen = new Set<string>();
  for (const id of selected) {
    if (!byId.has(id)) return 'Invalid nominee on ballot';
    if (seen.has(id)) return 'Duplicate selection on ballot';
    seen.add(id);
  }
  return null;
}

/** @deprecated Use validateElectionChoices. */
export function validateElectionRankings(
  options: PollOptionRow[],
  rankings: Record<string, Record<string, number>>,
  _committeeSeats?: number,
): string | null {
  void rankings;
  void _committeeSeats;
  if (options.length === 0) return 'No candidates';
  return 'Ranked ballots are no longer used. Refresh and cast a simple ballot.';
}

export type ElectedWinner = {
  option_id: string;
  name: string;
  score: number;
  place?: number;
  from_post?: ElectionPost;
};

export type FormationMember = {
  key: string;
  name: string;
  flat_number?: string | null;
  flat_id?: string | null;
  member_id?: string | null;
  option_id?: string | null;
  from_post?: ElectionPost | null;
  place?: number | null;
  source: 'voluntary' | 'executive_proposed' | 'runner_up';
};

export type CommitteeFormationState = {
  /** @deprecated Not used for new elections (bye-laws forbid auto-seating runners-up). */
  selected_runner_up_ids: string[];
  voluntary: FormationMember[];
  executive_proposed: FormationMember[];
};

export type ElectionResultsPayload = {
  president: ElectedWinner | null;
  vice_president: ElectedWinner | null;
  secretary: ElectedWinner | null;
  treasurer: ElectedWinner | null;
  committee: ElectedWinner[];
  /** Legacy only — never populated for new tallies. */
  runners_up?: Partial<Record<ElectionPost, ElectedWinner[]>>;
  vacant?: Partial<Record<ElectionPost, { reason: string; top_score: number; required: number }>>;
  formation?: CommitteeFormationState;
  tallied_at: string;
  ballot_mode?: 'combined' | 'separate_office' | 'ranked_legacy';
};

export function emptyFormationState(): CommitteeFormationState {
  return { selected_runner_up_ids: [], voluntary: [], executive_proposed: [] };
}

/** Legacy helper — returns [] for new tallies (runners_up never written). */
export function listRunnersUp(results: ElectionResultsPayload): ElectedWinner[] {
  const out: ElectedWinner[] = [];
  for (const post of OFFICE_POSTS) {
    for (const r of results.runners_up?.[post] ?? []) {
      out.push({ ...r, from_post: post });
    }
  }
  return out;
}

function scoreMapFromBallots(
  optionIds: string[],
  ballots: BallotRow[],
  markFor: (choices: BallotChoices, optionId: string) => boolean,
): Map<string, number> {
  const scores = new Map<string, number>();
  for (const id of optionIds) scores.set(id, 0);
  for (const b of ballots) {
    const choices = normalizeChoices(b.choices);
    for (const id of optionIds) {
      if (markFor(choices, id)) scores.set(id, (scores.get(id) ?? 0) + 1);
    }
  }
  return scores;
}

/** Legacy ranked → one vote for rank-1 only (migration path). */
function legacyRankOneScores(post: string, optionIds: string[], ballots: BallotRow[]): Map<string, number> {
  const scores = new Map<string, number>();
  for (const id of optionIds) scores.set(id, 0);
  for (const b of ballots) {
    if (b.choices && typeof b.choices === 'object' && Object.keys(normalizeChoices(b.choices)).some((k) => {
      const v = (normalizeChoices(b.choices) as Record<string, unknown>)[k];
      return Array.isArray(v) ? v.length > 0 : !!v;
    })) {
      continue;
    }
    const rmap = b.rankings?.[post] ?? {};
    for (const id of optionIds) {
      if (Number(rmap[id]) === 1) scores.set(id, (scores.get(id) ?? 0) + 1);
    }
  }
  return scores;
}

function rankedList(
  postOpts: PollOptionRow[],
  scores: Map<string, number>,
  fromPost: ElectionPost,
): ElectedWinner[] {
  return [...postOpts]
    .sort((a, b) => (scores.get(b.id) ?? 0) - (scores.get(a.id) ?? 0))
    .map((o, i) => ({
      option_id: o.id,
      name: o.option_text,
      score: scores.get(o.id) ?? 0,
      place: i + 1,
      from_post: fromPost,
    }));
}

export type TallyElectionOptions = {
  separateOfficeVotes?: boolean;
  committeeSeats?: number;
  /** @deprecated Ignored — runners-up are never auto-collected. */
  runnerUpPlaces?: number;
  /** @deprecated Ignored — Borda thresholds removed. */
  winningVotes?: Record<string, number>;
};

/**
 * Plurality / limited-vote tally from `choices`.
 * Does not auto-seat 2nd/3rd place. Does not use Borda ranking.
 */
export function tallyElection(
  options: PollOptionRow[],
  ballots: BallotRow[],
  committeeSeatsOrOpts: number | TallyElectionOptions = 3,
  winningVotesLegacy: Record<string, number> = {},
  runnerUpPlacesLegacy = 0,
): ElectionResultsPayload {
  const opts: TallyElectionOptions =
    typeof committeeSeatsOrOpts === 'number'
      ? {
          committeeSeats: committeeSeatsOrOpts,
          winningVotes: winningVotesLegacy,
          runnerUpPlaces: runnerUpPlacesLegacy,
        }
      : committeeSeatsOrOpts;

  const separate = Boolean(opts.separateOfficeVotes);
  const committeeSeats = Math.max(0, opts.committeeSeats ?? 3);
  const byPost = (post: string) => options.filter((o) => o.election_post === post);

  const hasAnyChoices = ballots.some((b) => {
    const c = normalizeChoices(b.choices);
    return (
      asStringArray(c.selected).length > 0 ||
      OFFICE_POSTS.some((p) => !!c[p]) ||
      asStringArray(c.committee).length > 0
    );
  });

  const ballotMode: ElectionResultsPayload['ballot_mode'] = hasAnyChoices
    ? separate
      ? 'separate_office'
      : 'combined'
    : 'ranked_legacy';

  const winners: Partial<Record<ElectionPost, ElectedWinner | null>> = {};

  if (separate || ballotMode === 'ranked_legacy') {
    for (const post of OFFICE_POSTS) {
      const postOpts = byPost(post);
      if (postOpts.length === 0) {
        winners[post] = null;
        continue;
      }
      const ids = postOpts.map((o) => o.id);
      const scores = hasAnyChoices
        ? scoreMapFromBallots(ids, ballots, (c, id) => c[post] === id)
        : legacyRankOneScores(post, ids, ballots);
      const ranked = rankedList(postOpts, scores, post);
      winners[post] = ranked[0] ?? null;
    }

    const committeeOpts = byPost('committee');
    const cIds = committeeOpts.map((o) => o.id);
    const cScores = hasAnyChoices
      ? scoreMapFromBallots(cIds, ballots, (c, id) => asStringArray(c.committee).includes(id))
      : legacyRankOneScores('committee', cIds, ballots);
    const committeeSorted = rankedList(committeeOpts, cScores, 'committee');
    const seats = Math.min(committeeSeats, committeeSorted.length);

    return {
      president: winners.president ?? null,
      vice_president: winners.vice_president ?? null,
      secretary: winners.secretary ?? null,
      treasurer: winners.treasurer ?? null,
      committee: committeeSorted.slice(0, seats),
      formation: emptyFormationState(),
      tallied_at: new Date().toISOString(),
      ballot_mode: ballotMode,
    };
  }

  // Combined: each marked nominee gets one mark; seat by nominated post.
  const allIds = options.map((o) => o.id);
  const scores = scoreMapFromBallots(allIds, ballots, (c, id) => asStringArray(c.selected).includes(id));

  for (const post of OFFICE_POSTS) {
    const postOpts = byPost(post);
    if (postOpts.length === 0) {
      winners[post] = null;
      continue;
    }
    winners[post] = rankedList(postOpts, scores, post)[0] ?? null;
  }

  const committeeOpts = byPost('committee');
  const committeeSorted = rankedList(committeeOpts, scores, 'committee');
  const seats = Math.min(committeeSeats, committeeSorted.length);

  return {
    president: winners.president ?? null,
    vice_president: winners.vice_president ?? null,
    secretary: winners.secretary ?? null,
    treasurer: winners.treasurer ?? null,
    committee: committeeSorted.slice(0, seats),
    formation: emptyFormationState(),
    tallied_at: new Date().toISOString(),
    ballot_mode: 'combined',
  };
}

File: mobile/lib/utils/election_tally.dart
import 'election_governance.dart';

class ElectedWinner {
  const ElectedWinner({
    required this.optionId,
    required this.name,
    required this.score,
    this.place,
    this.fromPost,
  });

  final String optionId;
  final String name;
  final int score;
  final int? place;
  final String? fromPost;

  Map<String, dynamic> toJson() => {
        'option_id': optionId,
        'name': name,
        'score': score,
        if (place != null) 'place': place,
        if (fromPost != null) 'from_post': fromPost,
      };
}

class ElectionResultsPayload {
  const ElectionResultsPayload({
    this.president,
    this.vicePresident,
    this.secretary,
    this.treasurer,
    this.committee = const [],
    this.formation = const {
      'selected_runner_up_ids': <String>[],
      'voluntary': <Map<String, dynamic>>[],
      'executive_proposed': <Map<String, dynamic>>[],
    },
    required this.talliedAt,
    this.ballotMode = 'combined',
  });

  final ElectedWinner? president;
  final ElectedWinner? vicePresident;
  final ElectedWinner? secretary;
  final ElectedWinner? treasurer;
  final List<ElectedWinner> committee;
  final Map<String, dynamic> formation;
  final String talliedAt;
  final String ballotMode;

  Map<String, dynamic> toJson() => {
        'president': president?.toJson(),
        'vice_president': vicePresident?.toJson(),
        'secretary': secretary?.toJson(),
        'treasurer': treasurer?.toJson(),
        'committee': committee.map((c) => c.toJson()).toList(),
        'formation': formation,
        'tallied_at': talliedAt,
        'ballot_mode': ballotMode,
      };
}

List<String> _asStringList(dynamic v) {
  if (v is List) return v.whereType<String>().where((s) => s.isNotEmpty).toList();
  if (v is String && v.isNotEmpty) return [v];
  return [];
}

Map<String, dynamic> _normalizeChoices(dynamic raw) {
  if (raw is! Map) return {};
  final c = Map<String, dynamic>.from(raw);
  return {
    'selected': _asStringList(c['selected']),
    'president': c['president'] is String ? c['president'] : null,
    'vice_president': c['vice_president'] is String ? c['vice_president'] : null,
    'secretary': c['secretary'] is String ? c['secretary'] : null,
    'treasurer': c['treasurer'] is String ? c['treasurer'] : null,
    'committee': _asStringList(c['committee']),
  };
}

bool _ballotHasChoices(Map<String, dynamic> choices) {
  if (_asStringList(choices['selected']).isNotEmpty) return true;
  for (final p in ['president', 'vice_president', 'secretary', 'treasurer']) {
    if (choices[p] is String && (choices[p] as String).isNotEmpty) return true;
  }
  return _asStringList(choices['committee']).isNotEmpty;
}

List<ElectedWinner> _rankedList(
  List<Map<String, dynamic>> postOpts,
  Map<String, int> scores,
  String fromPost,
) {
  final sorted = [...postOpts]
    ..sort((a, b) => (scores[b['id'] as String] ?? 0).compareTo(scores[a['id'] as String] ?? 0));
  return [
    for (var i = 0; i < sorted.length; i++)
      ElectedWinner(
        optionId: sorted[i]['id'] as String,
        name: sorted[i]['option_text'] as String? ?? '',
        score: scores[sorted[i]['id'] as String] ?? 0,
        place: i + 1,
        fromPost: fromPost,
      ),
  ];
}

/// Plurality / limited-vote tally from `choices` (not ranked/Borda).
ElectionResultsPayload tallyElection({
  required List<Map<String, dynamic>> options,
  required List<Map<String, dynamic>> ballots,
  required int committeeSeats,
  bool separateOfficeVotes = false,
  Map<String, int> winningVotes = const {},
  int runnerUpPlaces = 0,
}) {
  List<Map<String, dynamic>> byPost(String post) =>
      options.where((o) => o['election_post'] == post).toList();

  final normalized = ballots.map((b) => _normalizeChoices(b['choices'])).toList();
  final hasChoices = normalized.any(_ballotHasChoices);
  final separate = separateOfficeVotes;
  final seats = committeeSeats < 0 ? 0 : committeeSeats;

  Map<String, int> markScores(
    List<String> ids,
    bool Function(Map<String, dynamic> c, String id) mark,
  ) {
    final scores = {for (final id in ids) id: 0};
    for (final c in normalized) {
      for (final id in ids) {
        if (mark(c, id)) scores[id] = (scores[id] ?? 0) + 1;
      }
    }
    return scores;
  }

  ElectedWinner? winnerForOffice(String post) {
    final postOpts = byPost(post);
    if (postOpts.isEmpty) return null;
    final ids = postOpts.map((o) => o['id'] as String).toList();
    final scores = hasChoices
        ? (separate
            ? markScores(ids, (c, id) => c[post] == id)
            : markScores(ids, (c, id) => _asStringList(c['selected']).contains(id)))
        : <String, int>{for (final id in ids) id: 0};
    final ranked = _rankedList(postOpts, scores, post);
    return ranked.isEmpty ? null : ranked.first;
  }

  final committeeOpts = byPost('committee');
  final cIds = committeeOpts.map((o) => o['id'] as String).toList();
  final cScores = hasChoices
      ? (separate
          ? markScores(cIds, (c, id) => _asStringList(c['committee']).contains(id))
          : markScores(cIds, (c, id) => _asStringList(c['selected']).contains(id)))
      : <String, int>{for (final id in cIds) id: 0};
  final committeeSorted = _rankedList(committeeOpts, cScores, 'committee');
  final take = seats < committeeSorted.length ? seats : committeeSorted.length;

  return ElectionResultsPayload(
    president: winnerForOffice('president'),
    vicePresident: winnerForOffice('vice_president'),
    secretary: winnerForOffice('secretary'),
    treasurer: winnerForOffice('treasurer'),
    committee: committeeSorted.take(take).toList(),
    talliedAt: DateTime.now().toUtc().toIso8601String(),
    ballotMode: hasChoices ? (separate ? 'separate_office' : 'combined') : 'ranked_legacy',
  );
}

/// Legacy helper — new tallies never populate runners-up.
List<Map<String, dynamic>> listRunnersUpFromResults(dynamic results) {
  if (results is! Map) return [];
  final runners = results['runners_up'];
  if (runners is! Map) return [];
  final out = <Map<String, dynamic>>[];
  for (final post in threeExecutivePosts) {
    final list = runners[post];
    if (list is! List) continue;
    for (final r in list) {
      if (r is Map) {
        out.add({
          ...Map<String, dynamic>.from(r),
          'from_post': post,
        });
      }
    }
  }
  return out;
}

Map<String, dynamic> emptyFormationState() => {
      'selected_runner_up_ids': <String>[],
      'voluntary': <Map<String, dynamic>>[],
      'executive_proposed': <Map<String, dynamic>>[],
    };

Map<String, dynamic>? formationOf(dynamic results) {
  if (results is! Map) return null;
  final f = results['formation'];
  if (f is! Map) return emptyFormationState();
  return Map<String, dynamic>.from(f);
}

int countFormedCommitteeFromResults(dynamic results) {
  if (results is! Map) return 0;
  var n = 0;
  for (final key in ['president', 'vice_president', 'secretary', 'treasurer']) {
    if (results[key] is Map) n += 1;
  }
  final committee = results['committee'];
  if (committee is List) n += committee.length;
  final formation = formationOf(results);
  if (formation != null) {
    final vol = formation['voluntary'];
    final exec = formation['executive_proposed'];
    if (vol is List) n += vol.length;
    if (exec is List) n += exec.length;
  }
  return n;
}
